### PR TITLE
fix(core): hold `map_state` across the HAL map in `Buffer::map`

### LINKED ISSUE #9552: Race between buffer mapping and buffer destruction
There is a race between buffer mapping and buffer destruction (verified with a stress test, although it requires some extra debug checks in the code to identify the failure):
 
1. `Buffer::map` is called and does `mem::replace(&mut *self.map_state.lock(), BufferMapState::Idle)` to obtain the map state. The `map_state` lock is released immediately.  
2. `buffer_destroy` calls `unmap`, which sees `BufferMapState::Idle`, and allows the destruction to proceed. (`unmap` takes a read lock on the snatch guard, which does not prevent it from running concurrently with `Buffer::map`.)  
3. `Buffer::destroy` blocks waiting for a write lock on the snatch guard.
4. `Buffer::map` maps the buffer; eventually `maintain` finishes and drops the snatch guard  .
5. `Buffer::destroy` resumes, doesn’t find the buffer in any of the places that defer destruction, and destroys it while it is mapped.

In practice, `get_mapped_range` will be called before attempting to use the mapping, and will fail because the buffer is destroyed. But we shouldn't destroy mapped buffers.

### PR DESCRIPTION
**Pull request by AdrianEddy**
**Connections**

Fixes #9552
Supersedes #9610

**Description**

`Buffer::unmap` can fail with `BufferAccessError::NotMapped` for a buffer that is simply mid-map, when map and unmap run on different threads.

`Buffer::map` took the `map_state` lock as a temporary to swap the state out, ran the HAL map, then re-locked to store `Active`. Between those two acquisitions — which spans the entire HAL map — `map_state` is observably `Idle`. A concurrent `unmap` reads that `Idle` and lands in `unmap_inner`'s `Idle => Err(NotMapped)` arm. The `Active` arm had the same shape, so there were two observable-`Idle` windows.

The fix holds a single guard for the function. A concurrent unmap now blocks and then unmaps the `Active` mapping the call installs — a valid linearization, where `NotMapped` corresponded to no legal ordering at all.

The old comment warned that holding the lock across the match would deadlock "when we try to re-lock the buffer back to active". That re-lock is what this change removes — the stores reuse the guard already held.

Holding the guard across `device::map_buffer` also holds it across that function's `initialization_status` write, so `BUFFER_MAP_STATE` gains a `BUFFER_INITIALIZATION_STATUS` edge in `lock::rank`. The edge is acyclic: `initialization_status` is a leaf rank, and no path acquires `map_state` while holding a buffer's init tracker.

The third commit addresses a hazard the guard would otherwise introduce. `Device::handle_hal_error` calls `Device::lose`, which invokes the user's device-lost callback **inline**. On the HAL-map failure path that ran under the new guard, so a callback touching the same buffer (`unmap`, `destroy` → `unmap`, or `map_async`) would re-lock a non-reentrant mutex on the same thread; it also acquires `DEVICE_LOST_CLOSURE`, which is not a declared follower of `BUFFER_MAP_STATE`. `map_buffer_deferred_error` returns the HAL error untranslated so `Buffer::map` can release the guard first and translate after — `map_state` is already `Idle` there. `map_buffer` keeps its signature for the two buffer-creation call sites, which hold no such guard.

**Testing**

Reproduced on unmodified `trunk` with a stress harness using only the public API on the `noop` backend — thread A `map_async(Read)` then `poll(Wait)`, thread B one `unmap()` per iteration at swept phase offsets:

- trunk: **242 hits / 600k iterations**, every one with the impossible-under-atomic-semantics signature (unmap returns `NotMapped` while the same pending `mapAsync` resolves `Ok`), and every one confirmed by an immediate second `unmap()` succeeding with no intervening `map_async` — i.e. the state went `Idle → Active` with no legal writer.
- this PR: **0 hits / 600k**.

Under `--cfg wgpu_validate_locks`: trunk 178 hits / 200k, this PR 0 / 200k with no rank panics — the new edge is exercised on every successful map and the ranked validator accepts the guard discipline.

Those validator runs need #9960 applied on top; without it the mapping path aborts on a pre-existing release-order violation in `PendingSubmission` before reaching any of this. #9960 is independent of this PR and can land in either order.

`cargo xtask test buffer` — 330/331 on a discrete and an integrated GPU (real Vulkan and DX12 HAL maps), no hangs. The one failure, `wgpu_examples::big_compute_buffers::tests::two_buffers` on the integrated GPU, reproduces identically on unmodified `trunk`. `cargo test -p wgpu-core --all-features` — 65 pass.

The harness is not included here. It is ~130 lines of public API and could be trimmed into a regression test; happy to add it if wanted.

**Note for reviewers:** on GLES, `map_buffer` for `MAP_READ` does a `glGetBufferSubData` readback under the global context lock, so on that backend a concurrent same-buffer `unmap`/`map_async` can now block for the duration of a GPU sync rather than returning early. It is per-buffer and bounded, and the early return it replaces was a spurious error, but it is a real behavioural change on that backend.

**Squash or Rebase?**

Three commits, each independently reviewable; ready to rebase. Happy to squash if preferred.

**Checklist**

- [x] I self-reviewed and fully understand this PR.
- [ ] WebGPU implementations built with `wgpu` may be affected behaviorally.
- [ ] Validation and feature gates are in place to confine behavioral changes.
- [ ] Tests demonstrate the validation and altered logic works.
- [x] `CHANGELOG.md` entries for the user-facing effects of this change are present.
- [x] The PR is minimal, and doesn't make sense to land as multiple PRs.
- [x] Commits are logically scoped and individually reviewable.
- [x] The PR description has enough context to understand the motivation and solution implemented.

### CODE REVIEW

**Comment by ErichDonGubler** [comment]
#9960 is reviewed and mostly ready to merge. 🎉

**Comment by andyleiserson** [comment]
Related: #9552, #9610, #9479.

**Comment by andyleiserson** [comment]
Did you actually see deadlocks or other problems from calling the device loss callback while the lock was held, or is that logic included based on general correctness concerns? It would be better not to call it with locks held, but this not the only place we do, and I might prefer to address it in a more systematic way rather than adding different error wrappers/translators each time it comes up.

How long did it take to run 600k iterations with the test harness? As long as there isn't anything that makes the test actively CI-unfriendly (excessive resource consumption, tendency to provoke non-`wgpu` bugs, etc.), I'd be inclined to include it in CI with some modest iteration count (maybe 100ms worth, up to maybe 500ms or 1s worth if that could actually make the difference in whether it is a reliable regression test in CI.

**Comment by AdrianEddy** [comment]
Yes I've seen it in the wild while working on my app and stress-testing it. I've now included a new test which reproduces the issue. Without this PR, this new test fails. The test takes ~250ms on my PC

**Comment by AdrianEddy** [comment]
On the related issues:

- **#9552** — this PR fixes it, by the same mechanism that fixes the unmap race: with `map_state` held across the map, `destroy`'s internal `unmap`  blocks instead of observing the transient `Idle`, then unmaps the freshly installed mapping, so destruction can no longer proceed against a mid-map buffer. 
- **#9610** — same core idea as my first commit (hold `map_state` across the map, declare the `BUFFER_INITIALIZATION_STATUS` edge), so it fixes #9552 too. But it keeps the *translating* `map_buffer` under the held guard, which is exactly the deadlock in the test I added: `handle_hal_error` → `lose` → device-lost callback inline under `map_state` (hard hang in release, `DEVICE_LOST_CLOSURE after BUFFER_MAP_STATE` under `wgpu_validate_locks`). Its unit test also doesn't exercise `Buffer::map`/`unmap` — it asserts that a standalone mutex blocks a second locker, which holds on unfixed trunk too. So I believe this PR supersedes it: same fix, plus the deferred error translation and an interleaving-driven regression test.
- **#9479** — unmodified trunk already acquires `DEVICE_LOST_CLOSURE` under `QUEUE_LIFE_TRACKER` whenever a device is lost inside `poll`, so with validation on by default, any device loss during maintain panics in debug builds.

File: CHANGELOG.md [hunk 1]
```diff
@@ -126,6 +126,8 @@ By @sagudev in [#10115](https://github.com/gfx-rs/wgpu/pull/10115).
 - Fix required immediate slots calculation and remove `naga::valid::FunctionInfo::immediate_slots_used`. By @beicause in [#9725](https://github.com/gfx-rs/wgpu/pull/9725).
 - Fix a spurious assertion failure in `Device::maintain` when multiple threads race polling the same device. By @AdrianEddy in [#9958](https://github.com/gfx-rs/wgpu/pull/9958).
 - Fix `PendingSubmission` releasing its lock guards out of stacking order, which tripped `--cfg wgpu_validate_locks` on any submission. By @AdrianEddy in [#9960](https://github.com/gfx-rs/wgpu/pull/9960).
+- Fixed some cases where initialization tracking was not correct. By @andyleiserson in [#10002](https://github.com/gfx-rs/wgpu/pull/10002).
```
Comment (andyleiserson) [inline CHANGELOG.md]:
```suggestion
```
This seems like a conflict resolution error. This entry was extended to include a 2nd PR, the updated version appears a few lines below, so inserting it here is a duplication.

File: wgpu-core/src/device/mod.rs [hunk 2]
```diff
@@ -226,19 +226,52 @@ pub struct DeviceLostInvocation {
     message: String,
 }
 
+/// The failure modes of [`map_buffer_deferred_error`], kept apart so a caller
+/// can decide *when* to translate the HAL error.
```
Comment (andyleiserson) [inline wgpu-core/src/device/mod.rs]: I'd prefer to omit this special case unless there's a clear need to handle it here (actual problems observed due to calling the callback with locks held) -- this is not the only place we call `handle_hal_error` with locks held. One of the items in #10031 is to find a general solution for this, I don't think it's reasonable to add this amount of code in each place it happens.

File: wgpu-core/src/lock/rank.rs [hunk 3]
```diff
@@ -169,6 +169,10 @@ define_lock_ranks! {
         SHARED_TRACKER_INDEX_ALLOCATOR_INNER,
     }
     rank BUFFER_MAP_STATE "Buffer::map_state" followed by {
+        // `Buffer::map` holds `map_state` across `device::map_buffer`, which
+        // writes the buffer's initialization tracker. See `Buffer::map` for why
+        // the guard must span the HAL map.
```
Comment (andyleiserson) [inline wgpu-core/src/lock/rank.rs]:
```suggestion
```
I don't see why this edge is different from all the others in the graph. They could each have a comment explaining why the edge exists, but the comments would inevitably go stale and I don't think are worth it.

File: wgpu-core/src/resource.rs [hunk 4]
```diff
@@ -943,44 +943,60 @@ impl Buffer {
...
+        // duration of the HAL map below, so a concurrent `Buffer::unmap` saw
+        // `Idle` and failed with `NotMapped` for a buffer that was merely
+        // mid-map. Holding the guard makes that window unobservable: the unmap
+        // blocks here, then unmaps the `Active` mapping this call installs.
+        //
+        // The re-lock that previously required a temporary guard is gone; the
+        // stores below reuse this one, so there is nothing left to deadlock on.
+        //
+        // This also holds the guard across `crate::device::map_buffer`, which
+        // writes `initialization_status` — hence the new lock-rank edge.
```
Comment (andyleiserson) [inline wgpu-core/src/resource.rs]:
```suggestion
        // Hold the lock on `map_state` until we have updated it with the
        // outcome of the mapping, to prevent concurrent activity from
        // observing an intermediate state.
        //
        // Unfortunately this does mean that we hold the guard across
        // `crate::device::map_buffer` and the associated call to
        // `handle_hal_error`, which may invoke a device loss callback.
        // See <https://github.com/gfx-rs/wgpu/issues/10031>.
```

In most cases it's not useful for comments to describe the history of the code. Many corner cases I don't think are worth noting at all, tests are a better way of capturing corner cases and making sure we don't break them. But this one I do think is worth noting with a brief comment.

File: tests/tests/wgpu-validation/api/buffer_mapping.rs [hunk 5]
```diff
@@ -245,3 +245,131 @@ fn unmap_while_visible() {
...
+/// succeeds: the racing `unmap` may abort the pending mapping or unmap the
+/// freshly installed one, but it must not fail. Before the fix, `Buffer::map`
+/// swapped `map_state` out for the duration of the HAL map, and an `unmap`
+/// landing in that window failed with `NotMapped` even though its `map_async`
+/// went on to resolve `Ok` — an interleaving no sequential ordering of the two
+/// calls permits.
+///
+/// At the time of writing, unmodified trunk failed this test on every run,
+/// with ~700 spurious failures per 25 000 iterations and the first one inside
+/// the first dozen iterations. The full test takes ~0.25 s in a debug build.
```
Comment (andyleiserson) [inline tests/tests/wgpu-validation/api/buffer_mapping.rs]: The test takes ~1.7s on my Mac M1, so I was going to propose reducing the iteration count, but it also looks like the failure rate is much lower for me (single digits out of 25k), so that would suggest keeping it as-is. It would be worth updating the comment so somebody isn't tempted to reduce it in the future though.

File: tests/tests/wgpu-validation/api/buffer_mapping.rs [hunk 6]
```diff
@@ -245,3 +245,131 @@ fn unmap_while_visible() {
...
+///
+/// [#9959]: https://github.com/gfx-rs/wgpu/pull/9959
+#[test]
+#[cfg(not(target_arch = "wasm32"))]
+fn unmap_racing_map_never_spuriously_fails() {
```
Comment (andyleiserson) [inline tests/tests/wgpu-validation/api/buffer_mapping.rs]: nit: "never spuriously fails" doesn't seem like very informative wording in a test name. I might name the test `concurrent_mapping_and_unmapping`.

**Review by andyleiserson — CHANGES_REQUESTED** [review]
Sorry for the delay. I have enough comments that I'm going to request changes, but overall this is an important bug fix, if the comments are addressed we should be able to merge it promptly.

**Comment by AdrianEddy** [comment]
All feedback addressed

**Review by andyleiserson — APPROVED** [review]
Thanks!
